File: BhaktiChatAndroidNative/app/src/main/java/com/bhaktichat/app/domain/Aarti.kt
package com.bhaktichat.app.domain

enum class Deity {
    KRISHNA,
    GANESH,
    SHIV,
    LAKSHMI,
    DEVI,
    VISHNU,
    HANUMAN,
    OTHER
}

data class Aarti(
    val id: String,
    val slug: String,
    val title: String,
    val titleHi: String,
    val deity: Deity,
    val durationMinutes: Int?,
    val tags: List<String>,
    val youtubeVideoId: String?,
    val audioUrl: String?,
    val popularityCount: Long?,
    val isTop: Boolean,
    val lyrics: List<String>,
    /** Per-aarti generated artwork (drawable resource name), when this specific aarti has its
     * own image rather than sharing a generic per-deity fallback. */
    val imageAssetName: String?
) {
    /** True when a real recorded aarti track is available to stream (vs. lyrics-only / TTS). */
    val hasAudio: Boolean get() = !audioUrl.isNullOrBlank()

    val subtitle: String?
        get() = titleHi.takeIf { it.isNotBlank() && !it.equals(title, ignoreCase = true) }

    val preview: String
        get() = lyrics
            .filter { it.isNotBlank() && !it.endsWith("...") }
            .take(2)
            .joinToString(" ")
            .ifBlank { "पूरे बोल पढ़ने के लिए टैप करें।" }
}

File: BhaktiChatiOS/Sources/Core/Models/Aarti.swift
import Foundation

enum Deity: String, Codable {
    case krishna = "KRISHNA"
    case ganesh = "GANESH"
    case shiv = "SHIV"
    case lakshmi = "LAKSHMI"
    case devi = "DEVI"
    case vishnu = "VISHNU"
    case hanuman = "HANUMAN"
    case other = "OTHER"
}

struct Aarti: Identifiable, Codable, Hashable {
    let id: String
    let slug: String
    let title: String
    let titleHi: String
    let deity: Deity
    let durationMinutes: Int?
    let tags: [String]
    let youtubeVideoId: String?
    let popularityCount: Int?
    let isTop: Bool
    let lyrics: [String]
    let audioUrl: String?
    /// Per-aarti generated artwork (PackageAssetLoader asset name), when this specific aarti
    /// has its own image rather than sharing a generic per-deity fallback.
    let imageAsset: String?

    var hasAudio: Bool {
        guard let audioUrl else { return false }
        return !audioUrl.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    var subtitle: String? {
        titleHi
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .takeIf { !$0.isEmpty && $0.caseInsensitiveCompare(title) != .orderedSame }
    }

    var preview: String {
        let lines = lyrics
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty && !$0.hasSuffix("...") }
            .prefix(2)

        let text = lines.joined(separator: " ")
        return text.isEmpty ? "Tap to read full lyrics." : text
    }

    var lyricVerses: [String] {
        lyrics
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
    }

    var hasArtwork: Bool {
        if imageAsset != nil { return true }
        switch deity {
        case .ganesh, .shiv, .lakshmi, .krishna, .hanuman:
            return true
        default:
            return false
        }
    }
}

private extension String {
    func takeIf(_ predicate: (String) -> Bool) -> String? {
        predicate(self) ? self : nil
    }
}
